include/nbl/builtin/hlsl/tgmath.hlsl

```diff
@@ -4,70 +4,117 @@
 #ifndef _NBL_BUILTIN_HLSL_TGMATH_INCLUDED_
 #define _NBL_BUILTIN_HLSL_TGMATH_INCLUDED_
 
+#include <nbl/builtin/hlsl/spirv_intrinsics/glsl.std.450.hlsl>
+#include <nbl/builtin/hlsl/impl/tgmath_impl.hlsl>
+#include <nbl/builtin/hlsl/type_traits.hlsl>
+#include <nbl/builtin/hlsl/vector_utils/vector_traits.hlsl>
 #include <nbl/builtin/hlsl/cpp_compat.hlsl>
 #include <nbl/builtin/hlsl/ieee754.hlsl>
-#include <nbl/builtin/hlsl/type_traits.hlsl>
-
 #include <nbl/builtin/hlsl/spirv_intrinsics/core.hlsl>
+#include <nbl/builtin/hlsl/concepts/core.hlsl>
+#include <nbl/builtin/hlsl/concepts/vector.hlsl>
+// C++ headers
+#ifndef __HLSL_VERSION
+#include <algorithm>
+#include <cmath>
+#endif
 
 namespace nbl
 {
 namespace hlsl
 {
-namespace tgmath
-{
-
-//template <typename T>
-//inline bool isNaN(T val)
-//{
-//	using AsUint = typename unsigned_integer_of_size<sizeof(T)>::type;
-//	using AsFloat = typename float_of_size<sizeof(T)>::type;
-//
-//	AsUint asUint = bit_cast<AsUint, T>(val);
-//	return bool((ieee754::extractBiasedExponent<T>(val) == ieee754::traits<AsFloat>::specialValueExp) && (asUint & ieee754::traits<AsFloat>::mantissaMask));
-//}
-//
-//template<typename T>
-//inline bool isInf(T val)
-//{
-//	using AsUint = typename unsigned_integer_of_size<sizeof(T)>::type;
-//	using AsFloat = typename float_of_size<sizeof(T)>::type;
-//
-//	AsUint tmp = bit_cast<AsUint>(val);
-//	return (tmp & (~ieee754::traits<AsFloat>::signMask)) == ieee754::traits<AsFloat>::inf;
-//}
-//
-//#ifdef __HLSL_VERSION
-//#define DEFINE_IS_NAN_SPECIALIZATION(TYPE)\
-//template<>\
-//inline bool isNaN<TYPE>(TYPE val)\
-//{\
-//	return spirv::isNan(val);\
-//}\
-//
-//#define DEFINE_IS_INF_SPECIALIZATION(TYPE)\
-//template<>\
-//inline bool isInf<TYPE>(TYPE val)\
-//{\
-//	return spirv::isInf(val);\
-//}\
-//
-//DEFINE_IS_NAN_SPECIALIZATION(float16_t)
-//DEFINE_IS_NAN_SPECIALIZATION(float32_t)
-//DEFINE_IS_NAN_SPECIALIZATION(float64_t)
-//
-//DEFINE_IS_INF_SPECIALIZATION(float16_t)
-//DEFINE_IS_INF_SPECIALIZATION(float32_t)
-//DEFINE_IS_INF_SPECIALIZATION(float64_t)
-//
-//#undef DEFINE_IS_INF_SPECIALIZATION
-//#undef DEFINE_IS_NAN_SPECIALIZATION
-//#undef INTRINSIC_FUNC_NAMESPACE
-//#endif
-
-}
-
-}
-}
-
-#endif
\ No newline at end of file
+// TODO: will not work for emulated_float as an input because `concepts::floating_point<T>` is only for native floats, fix every occurance
+template<typename T>
+inline T erf(T x)
+{
+    return tgmath_impl::erf_helper<T>::__call(x);
+}
+
+template<typename T>
+inline T erfInv(T x)
+{
+    return tgmath_impl::erfInv_helper<T>::__call(x);
+}
+
+template<typename T>
+inline T floor(NBL_CONST_REF_ARG(T) val)
+{
+    return tgmath_impl::floor_helper<T>::__call(val);
+}
+
+template<typename T, typename U>
+inline T lerp(NBL_CONST_REF_ARG(T) x, NBL_CONST_REF_ARG(T) y, NBL_CONST_REF_ARG(U) a)
+{
+    return tgmath_impl::lerp_helper<T, U>::__call(x, y, a);
+}
```

Comment:
I'd prefer sticking to GLSL-ism and only having a `mix` and then `lerp` solely as a top level alias